Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='green'><b>Полезные (и просто интересные) материалы:</b></font> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .
</font>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Отлично, что стоп-слова были исключены при векторизации!
* Работа получилась отличной, тебе удалось добиться достаточно хорошего качества. Поздравляю!
* Проект может быть зачтен, но я его отправлю назад, чтобы у тебя была возможность задать вопросы и внести правки, при желании. Однако, ты можешь просто вернуть проект в таком же виде и я его зачту.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

In [1]:
pip install nltk

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import nltk
import re
import numpy as np
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from lightgbm import LGBMModel,LGBMClassifier
from sklearn.svm import LinearSVC
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.model_selection import cross_val_score
from pymystem3 import Mystem
from nltk.stem import WordNetLemmatizer
from sklearn.metrics import mean_squared_error
import warnings
warnings.filterwarnings("ignore")

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [3]:
df = pd.read_csv('/datasets/toxic_comments.csv')

In [4]:
df.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
text     159571 non-null object
toxic    159571 non-null int64
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [6]:
df.duplicated().sum()

0

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корректно, первичный осмотр проведен.
</div>

Напишем функцию лемматизации текста:

In [7]:
lemmatizer = WordNetLemmatizer()

In [8]:
def lemmatize(text):
    words = nltk.word_tokenize(text)
    lemmatized_output = ' '.join([lemmatizer.lemmatize(w) for w in words])
    
    return lemmatized_output

Напишем функцию очистки текста от лишних символов:

In [9]:
def clear_text(text):
    text = re.sub(r'[^a-zA-Z ]', ' ', text)
    
    return " ".join(text.split())

Приведем текст к кодировке Unicode:

In [10]:
corpus = df['text'].values.astype('U')

Проведем лемматизацию и очистку текста:

In [11]:
corpus_lemm = [lemmatize(clear_text(corpus[i])) for i in range(len(corpus))]

<div class="alert alert-block alert-success">
<b>Успех:</b> Очистка и лемматизация были сделаны правильно. Радует, что ты их применил именно в таком порядке.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Можно было напечатать несколько примеров до/после.
</div>

Загрузим стоп-слова:

In [12]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Разобъем данные на выборки и очистим их от стоп-слов:

In [13]:
X_train, X_test, y_train, y_test = train_test_split(corpus_lemm, df['toxic'],
                                                    train_size=0.8,
                                                    test_size=0.2,
                                                    random_state=123)

Применим TF-IDF:

In [14]:
tf_idf_model = TfidfVectorizer(stop_words=stopwords)
tf_idf_model.fit_transform(X_train)
res_test = tf_idf_model.transform(X_test)
res_train = tf_idf_model.transform(X_train)

<b>Вывод</b>

Данные содержат 159571 строк и 2 столбца: комментарий и признак его токсичности. Дубликатов в данных не обнаружено.  
Для того, чтобы получить корректные результаты, выполнена подготовка текста: выделены леммы каждого слова, текст очищен от лишних символов и приведен к кодировке Unicode.  
Выполнено разбиение данных на обучающую и тестовую выборку, создана матрица cо значениями TF-IDF по корпусу комментариев.

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что векторизатор был обучен только на тренировочной части данных. Это уменьшает переобучение.
</div>

## Обучение

<b>Логистическая регрессия</b>

In [15]:
parameters = {'C': np.linspace(10, 20, num = 11, endpoint = True),
             'max_iter': [1000]}
lrm = LogisticRegression()
gscv_lrm = GridSearchCV(lrm, parameters,
                  cv=5,
                  scoring='f1')
gscv_lrm.fit(res_train, y_train)

print(f"Наилучшая оценка f1: {gscv_lrm.best_score_:.3f}")

Наилучшая оценка f1: 0.766


<div class="alert alert-block alert-info">
<b>Совет: </b> Напомню, что внутри кросс-валидации происходих разбиение выборки на треин и валидацию. Однако, в таком случае векторизатор обучен на всей выборке, а это не совсем корректно. Для избежания такого эффекта можно использовать <a href="https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">пайплайн</a>.
</div>

<b>Дерево решений</b>

In [16]:
parameters = {'max_depth': [5, 10, 15]}
dt = DecisionTreeClassifier()
gscv_dt = GridSearchCV(dt, parameters, cv=5, scoring='f1')
gscv_dt.fit(res_train, y_train)

print(f"Наилучшая оценка f1: {gscv_dt.best_score_:.3f}")

Наилучшая оценка f1: 0.617


<b>LinearSVC</b>

In [18]:
parameters = {'C': np.linspace(1, 31, num = 7, endpoint = True)}
lsvcm = LinearSVC(max_iter = 1000)
gscv_lsvc = GridSearchCV(lsvcm, parameters,
                  cv=5,
                  scoring='f1')
gscv_lsvc.fit(res_train, y_train)

print(f"Наилучшая оценка f1: {gscv_lsvc.best_score_:.3f}")

Наилучшая оценка f1: 0.771


<b>Вывод</b>

Наилучшие результаты показала модель LinearSVC, проверим ее на тестовой выборке:

In [20]:
best_model = gscv_lsvc.best_estimator_
predict = best_model.predict(res_test)
f1_lsvc = f1_score(y_test, predict)
print('f1 наилучшей модеди на тестовой выборке: ', f1_lsvc)

f1 наилучшей модеди на тестовой выборке:  0.7748856841364755


<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что попробовал несколько разных моделей!
</div>

## Выводы

Осуществлена обработка данных, содержащих токсичные комментарии:
- выполнена лемматизация,
- удалены лишние знаки,
- удалены стоп-словаСтоп слова.
Векторизация корпуса выполнена с помощью TfidfVectorizer.
Были обучены следующие виды моделей: LogisticRegression, DecisionTreeClassifier и LinearSVC. 
Наибольшую метрику f1 удалось получить на LinearSVC. На тестовых данных f1 составила 0.77, что удовлетворяет требованию технического задания.

<div class="alert alert-block alert-success">
<b>Успех:</b> Приятно видеть хорошо структурированынй вывод в конце проекта!
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны